common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3801,6 +3799,466 @@ static jint NativeCrypto_EVP_AEAD_CTX_open_buf(JNIEnv* env, jclass, jlong evpAea
+        return nullptr;
+    }
+    const EVP_HPKE_AEAD* aead = getHpkeAead(env, aeadValue);
+    if (aead == nullptr) {
+        return nullptr;
+    }
+    if (publicKeyArray == nullptr || seedArray == nullptr) {
+        conscrypt::jniutil::throwNullPointerException(env, "publicKeyArray or seedArray == null");
+        return {};
+    }
+
+    jbyte* peer_public_key = env->GetByteArrayElements(publicKeyArray, 0);
+
+    jbyte* seed = env->GetByteArrayElements(seedArray, 0);
+
+    jbyte* infoArrayElement = nullptr;
+    const uint8_t* info = nullptr;
+    size_t infoLen = 0;
+    if (infoArray != nullptr) {
+        infoArrayElement = env->GetByteArrayElements(infoArray, 0);
+        info = reinterpret_cast<const uint8_t*>(infoArrayElement);
+        infoLen = env->GetArrayLength(infoArray);
+    }
+
+    size_t encapsulatedSharedSecretLen;
+    std::vector<uint8_t> encapsulatedSharedSecret(EVP_HPKE_MAX_ENC_LENGTH);
+
+    bool isSetup = EVP_HPKE_CTX_setup_sender_with_seed_for_testing(
+            /* ctx= */ ctx,
+            /* out_enc= */ encapsulatedSharedSecret.data(),
+            /* out_enc_len= */ &encapsulatedSharedSecretLen,
+            /* max_enc= */ encapsulatedSharedSecret.size(),
+            /* kem= */ kem,
+            /* kdf= */ kdf,
+            /* aead= */ aead,
+            /* peer_public_key= */ reinterpret_cast<const uint8_t*>(peer_public_key),
+            /* peer_public_key_len= */ env->GetArrayLength(publicKeyArray),
+            /* info= */ info,
+            /* info_len= */ infoLen,
+            /* seed= */ reinterpret_cast<const uint8_t*>(seed),
+            /* seed_len= */ env->GetArrayLength(seedArray));
+
+    env->ReleaseByteArrayElements(publicKeyArray, peer_public_key, JNI_ABORT);
+    env->ReleaseByteArrayElements(seedArray, seed, JNI_ABORT);
+    if (infoArrayElement != nullptr) {
+        env->ReleaseByteArrayElements(infoArray, infoArrayElement, JNI_ABORT);
+    }
+
+    if (!isSetup) {
+        conscrypt::jniutil::throwExceptionFromBoringSSLError(
+                env, "EVP_HPKE_CTX_setup_sender_with_seed_for_testing");
+        return {};
+    }
+
+    jbyteArray encArray = env->NewByteArray(encapsulatedSharedSecretLen);
+    env->SetByteArrayRegion(encArray, 0, encapsulatedSharedSecretLen,
+                            reinterpret_cast<const jbyte*>(encapsulatedSharedSecret.data()));
+
+    ScopedLocalRef<jobjectArray> result(
+            env, env->NewObjectArray(2, conscrypt::jniutil::objectClass, nullptr));
+
+    jobject ctxObject = env->NewObject(conscrypt::jniutil::nativeRefHpkeCtxClass,
+                                       conscrypt::jniutil::nativeRefHpkeCtxClass_constructor,
+                                       reinterpret_cast<jlong>(ctx));
+    env->SetObjectArrayElement(result.get(), 0, ctxObject);
+    env->SetObjectArrayElement(result.get(), 1, encArray);
+
+    return result.release();
+}
+
+static void NativeCrypto_EVP_HPKE_KEY_free(JNIEnv* env, jclass, jlong hpkeKeyRef) {
```

Comment:
Is this, and the corresponding Java bits, ever used? It didn't look like your API ever exposed `EVP_HPKE_KEY`.

(If a single recipient is decrypting multiple messages, using an `EVP_HPKE_KEY` over the underlying bytes is preferable because it saves you a base point multiplication on every decryption.)

Reply:
It was previously used, but after refactoring in an earlier commit I missed removing this. I have removed this method now. Thanks for pointing it out!